### PR TITLE
Refactor registration_controller.rb

### LINKED ISSUE #245: Refactor how registrations_controller handles validations of incoming request information
- [x] change user_impersonation to insufficient permissions 
- [x] remove user_impersonation from error_codes
- [x] add a NOTE about why events can only be empty when status is deleted (to be converted to documentation)

### PR DESCRIPTION
**Pull request by dunkOnIT**
This PR introduces a number of refactors intended to reduce duplication, make the code simpler, and make some other minor improvements we've been speaking about. 

Resolves #245

### CODE REVIEW

**Comment by FinnIckler** [comment]
Any way we can get this rebased ;)?

**Comment by dunkOnIT** [comment]
Lol yes my soul left my body once I came to the PR. Will get it done on Sunday :D

**Comment by dunkOnIT** [comment]
@FinnIckler the diff is now accurate. Do we care about trying to resolve the long commits?

**Comment by dunkOnIT** [comment]
- [ ] Look into disabling rubocop super check only for RegistrationError

File: app/controllers/registration_controller.rb [hunk 1]
```diff
@@ -245,46 +187,46 @@ def payment_ticket
...
 
-  def list
-    competition_id = list_params
-    competition_info = CompetitionApi.get_competition_info(competition_id)
-
-    if competition_info[:competition_exists?] == false
-      return render json: { error: competition_info[:error] }, status: competition_info[:status]
+  def validate_payment_ticket_request
+    competition_id = params[:competition_id]
+    unless CompetitionApi.uses_wca_payment?(competition_id)
```
Comment (dunkOnIT) [inline app/controllers/registration_controller.rb:192]: combine

**Comment by dunkOnIT** [comment]
- [ ] In validate_create_request, have a `user_can_create_registration` method that aliases to shared logic
- [ ] Considering refactoring the "only admins can register if comp is open (put that in the `user_can_create_registration`) 
- [ ] do raise/unless on one line
- [ ] Ask ChatGPT about rewriting if params.key?(:competing)

**Review by FinnIckler — APPROVED** [review]
Just what we talked about in the call

**Pull request merged**